configuration/RealTimeTranslation/webResourceForCCaaS.js

```diff
@@ -0,0 +1,714 @@
+	getISO6391LanguageCodeFromOcLanguageCode: function (ocLanguageCode) {
+		var ISO6391LanCode = "invalid code"; //invalid code
+		ISO6391LanCode = "en";
+		for (var key in C1WebResourceNamespace.ISO6391LanguageCodeToOcLanguageCodeMap) {
+			if (C1WebResourceNamespace.ISO6391LanguageCodeToOcLanguageCodeMap[key].includes(ocLanguageCode)) {
+				ISO6391LanCode = key;
+				break;
+			}
+		}
+		return ISO6391LanCode;
+	},
+
+	//saves the c2 langauge to CDS's contextVariable table for the given conversation.
+	//So, that it is remembered across multiple transfers for same conversation
+	//i.e. it is shared with multiple agents during transfer of conversation
+	upsertC2LanguageInCRM: function (conversationId, c2Lang, token, crmUrl) {
+		var data = {
+			"msdyn_value": c2Lang,
+			"msdyn_name": "msdyn_C2_language",
+			"statecode": 0,
+			"statuscode": 1,
+			"msdyn_isdisplayable": true,
+			"msdyn_ocliveworkitemid@odata.bind": "/msdyn_ocliveworkitems(" + conversationId + ")"
+		};
+		consoleLogHelper(conversationId, "Before upserting C2 Language in CRM", {
+			c2Lang,
+			dictionary: C1WebResourceNamespace.dictForAllConversation[conversationId],
+			data
+		});
+		if (C1WebResourceNamespace.dictForAllConversation[conversationId]['msdyn_C2_language_id'] == null) {
+			// create record
+			C1WebResourceNamespace.createXHRRequestPromise("POST", crmUrl + "/api/data/v9.0/msdyn_ocliveworkitemcontextitems", data, token).then(
+				function success(result) {
+					consoleLogHelper(conversationId, "created msdyn_ocliveworkitemcontextitem when msdyn_C2_language_id is null", {
+						result
+					});
+					C1WebResourceNamespace.dictForAllConversation[conversationId]['msdyn_C2_language_id'] = String(result.id);
+				},
+				function (error) {
+					consoleLogHelper(conversationId, "Failed to create msdyn_ocliveworkitemcontextitem", error, true);
+					// handle error conditions
+				}
+			);
+		} else {
+			// update record
+			C1WebResourceNamespace.createXHRRequestPromise("PATCH", crmUrl + "/api/data/v9.0/msdyn_ocliveworkitemcontextitems(" + C1WebResourceNamespace.dictForAllConversation[conversationId]["msdyn_C2_language_id"] + ")", data, token).then(
+				function success(result) {
+					consoleLogHelper(conversationId, "updated msdyn_ocliveworkitemcontextitem", {
+						result
+					});
+				},
+				function (error) {
+					consoleLogHelper(conversationId, "Failed to update msdyn_ocliveworkitemcontextitem", error, true);
+					// handle error conditions
+				}
+			);
+		}
+	},
+
+	//This method constructs and execute the XHR request
+	createXHRRequestPromise: function (method, url, data, token) {
+		return new Promise(function (resolve, reject) {
+			var req = new XMLHttpRequest();
+			req.open(method, url, true);
+			req.setRequestHeader("Accept", "application/json");
+			req.setRequestHeader("Content-Type", "application/json");
+			req.setRequestHeader("Authorization", `Bearer ${token}`);

```

Comment:
what if token is null here?

Reply:
If the token is null, a 401 status will be returned, and the `onreadystatechange()` function will handle it by rejecting the request.